src/utils/recorder_events.jl

```diff
@@ -0,0 +1,290 @@
+abstract type AbstractRecorderEvent end
+
+get_common(event::AbstractRecorderEvent) = event.common
+get_name(event::AbstractRecorderEvent) = get_name(get_common(event))
+get_timestamp(event::AbstractRecorderEvent) = get_timestamp(get_common(event))
+
+struct RecorderEventCommon
+    name::String
+    timestamp::Dates.DateTime
+end
+
+function RecorderEventCommon(name::AbstractString)
+    return RecorderEventCommon(name, Dates.now())
+end
+
+get_name(event::RecorderEventCommon) = event.name
+get_timestamp(event::RecorderEventCommon) = event.timestamp
+
+function serialize(event::T) where {T <: AbstractRecorderEvent}
+    data = Dict{Symbol, Any}()
+
+    for (fieldname, fieldtype) in zip(fieldnames(T), fieldtypes(T))
+        if fieldtype <: RecorderEventCommon
+            data[:name] = get_name(event)
+            data[:timestamp] = get_timestamp(event)
+        else
+            data[fieldname] = getfield(event, fieldname)
+        end
+    end
+
+    return JSON.json(data)
+end
+
+"""
+    deserialize(::Type{T}, text::AbstractString) where T <: AbstractRecorderEvent
+
+Deserialize an event from raw text. Will throw an exception if the text does not contain
+a serialized version of T.
+"""
+function deserialize(::Type{T}, text::AbstractString) where {T <: AbstractRecorderEvent}
+    data = JSON.parse(text)
+    name = pop!(data, "name")
+    timestamp = Dates.DateTime(pop!(data, "timestamp"))
+    common = RecorderEventCommon(name, timestamp)
+    vals = []
+
+    for (fieldname, fieldtype) in zip(fieldnames(T), fieldtypes(T))
+        if fieldtype <: RecorderEventCommon
+            push!(vals, common)
+        else
+            push!(vals, data[string(fieldname)])
+        end
+    end
+
+    return T(vals...)
+end
+
+"""
+Records user-defined events in JSON format.
+"""
+mutable struct Recorder
+    name::Symbol
+    io::IO
```

Comment:
I considered storing a subtype of Logging.AbstractLogger instead of an IO.  I decided against that because we don't need the benefits of the logger here.
- Controlled whether a particular recorder is enabled or not is done with the register_recorder! function.  The "log level" will always be the same.
- The default logger adds things like file and line number to the message automatically.  We would have to write code to prevent that.

It would be easy to switch to a logger if you have arguments for it.